src/execution/window_segment_tree.cpp

```diff
@@ -58,20 +61,21 @@ unique_ptr<WindowAggregatorState> WindowAggregator::GetGlobalState(idx_t group_c
 	return make_uniq<WindowAggregatorGlobalState>(*this, group_count);
 }
 
-void WindowAggregator::Sink(WindowAggregatorState &gsink, DataChunk &payload_chunk, SelectionVector *filter_sel,
-                            idx_t filtered) {
+void WindowAggregator::Sink(WindowAggregatorState &gsink, DataChunk &payload_chunk, idx_t input_idx,
+                            SelectionVector *filter_sel, idx_t filtered) {
 	auto &gasink = gsink.Cast<WindowAggregatorGlobalState>();
-	auto &inputs = gasink.inputs;
+	auto &winputs = gasink.winputs;
 	auto &filter_mask = gasink.filter_mask;
-	auto &filter_pos = gasink.filter_pos;
-	if (inputs.ColumnCount()) {
-		inputs.Append(payload_chunk, true);
+	if (winputs.chunk.ColumnCount()) {
+		winputs.Copy(payload_chunk, input_idx);
 	}
 	if (filter_sel) {
+		//	Single threaded for now.
+		//	TODO: Check for mask boundaries.
+		lock_guard<mutex> sink_guard(gasink.lock);
 		for (idx_t f = 0; f < filtered; ++f) {
-			filter_mask.SetValid(filter_pos + filter_sel->get_index(f));
+			filter_mask.SetValid(input_idx + filter_sel->get_index(f));
```

Comment:
Maybe it makes sense to turn `filter_mask` into a `bool[]` array as well to allow for concurrent access?

Reply:
I think that would work. It might be worth wrapping it so the calls look the same.